tritium-jmh/src/jmh/java/com/palantir/tritium/microbenchmarks/ReservoirBenchmarks.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.microbenchmarks;
+
+import com.codahale.metrics.ExponentiallyDecayingReservoir;
+import com.codahale.metrics.Reservoir;
+import com.palantir.tritium.metrics.registry.LockFreeExponentiallyDecayingReservoir;
+import java.util.concurrent.TimeUnit;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(1)
+@Threads(32)
+@State(Scope.Benchmark)
+@SuppressWarnings({"designforextension", "NullAway"})
+public class ReservoirBenchmarks {
+
+    @Param({"EXPO_DECAY", "LOCK_FREE_EXPO_DECAY"})
+    private ReservoirType reservoirType;
+
+    public enum ReservoirType {
+        EXPO_DECAY() {
+            @Override
+            Reservoir create() {
+                return new ExponentiallyDecayingReservoir();
+            }
+        },
+        LOCK_FREE_EXPO_DECAY() {
+            @Override
+            Reservoir create() {
+                return new LockFreeExponentiallyDecayingReservoir();
+            }
+        };
+
+        abstract Reservoir create();
+    }
+
+    private Reservoir reservoir;
+
+    @Setup
+    public void before() {
+        reservoir = reservoirType.create();
+    }
+
+    @Benchmark
+    public void updateReservoir() {
+        reservoir.update(1L);
```

Comment:
Is it worth changing the rescale time to test rescale performance as well?

Reply:
It's difficult to compare to the upstream version because they do not allow the interval to be configured. I think it's infrequent enough that we shouldn't need to worry about it.

Reply:
+1 for benchmarks :)

The upstream dropwizard-metrics implementation allows passing in a clock - could you inject a mutable clock which you increment by a second or so after every update? 

```
public ExponentiallyDecayingReservoir(int size, double alpha, Clock clock) {
```

or perhaps you'd even put a for loop in this @Benchmark method to do exactly 3600 updates and guarantee you get a rescale on every invocation?

Reply:
We can increment a clock, but things get a little wonky because we're running on several threads simultaneously. I think it boils down to what data we want to distill from the benchmark. The worst case scenario for this implementation is when all N threads concurrently do the work to rescale rather than parking while a single thread holding a lock does the work -- in the single-thread case the rescale itself is equivalent.

Relatedly, if we're worried about threads which produce metric running into issues, we could update this implementation such that it only rescales on `getSnapshot()`, but I'm not confident it's a good idea as long as the rescale interval is somewhat configurable.